internal/core/analysis/policy_engine.go

```diff
@@ -0,0 +1,162 @@
+		return testResults, err
+	}
+	engineInput := enginemodels.PolicyEngineInput{
+		Policies: []enginemodels.Policy{
+			{
+				Body:          string(policy.Body),
+				ID:            testPolicyID, // doesn't matter, we're only running one policy
+				ResourceTypes: policy.ResourceTypes,
+			},
+		},
+		Resources: testResources,
+	}
+
+	engineOutput, err := e.Run(engineInput)
+	if err != nil {
+		return testResults, errors.Wrap(err, "error calling policy engine")
+	}
+
+	// Translate policy engine output to test results.
+	for _, result := range engineOutput.Resources {
+		// Determine which test case this result corresponds to. We constructed resourceID with the
+		// format Panther:Test:Resource:TestNumber (see testResourceID),
+		testIndex, err := strconv.Atoi(strings.Split(result.ID, ":")[3])
+		if err != nil {
+			return testResults, errors.Wrapf(err, "unable to extract test number from test result resourceID %s", result.ID)
+		}
+
+		test := policy.Tests[testIndex]
+		switch {
+		case len(result.Errored) > 0:
+			// There was an error running this test, store the error message
+			testResults.TestsErrored = append(testResults.TestsErrored, &models.TestErrorResult{
+				ErrorMessage: result.Errored[0].Message,
+				Name:         string(test.Name),
+			})
+			testResults.TestSummary = false
+
+		case len(result.Failed) > 0 && bool(test.ExpectedResult), len(result.Passed) > 0 && !bool(test.ExpectedResult):
+			// The test result was not expected, so this test failed
+			testResults.TestsFailed = append(testResults.TestsFailed, string(test.Name))
+			testResults.TestSummary = false
+
+		case len(result.Failed) > 0 && !bool(test.ExpectedResult), len(result.Passed) > 0 && bool(test.ExpectedResult):
+			// The test result was as expected
+			testResults.TestsPassed = append(testResults.TestsPassed, string(test.Name))
+			testResults.TestSummary = true
+
+		default:
+			// This test didn't run (result.{Errored, Failed, Passed} are all empty). This must not happen absent a bug.
+			return testResults, errors.Errorf("unable to run test for resourceID %s", result.ID)
+		}
+	}
+	return testResults, nil
+}
+
+func (e *PolicyEngine) Run(input enginemodels.PolicyEngineInput) (enginemodels.PolicyEngineOutput, error) {
+	var engineOutput enginemodels.PolicyEngineOutput
+
+	payload, err := jsoniter.Marshal(&input)
+	if err != nil {
+		return engineOutput, errors.Wrap(err, "failed to marshal PolicyEngineInput")
+	}
+
+	invokeOutput, err := e.lambdaClient.Invoke(&lambda.InvokeInput{FunctionName: &e.lambdaName, Payload: payload})
+	if err != nil {
+		return engineOutput, errors.Wrap(err, "failed to invoke the policy engine")
+	}
+	if invokeOutput.FunctionError != nil {
+		return engineOutput, errors.Errorf("failed to run policy:\n %s", *invokeOutput.FunctionError)
+	}
+
+	if err := jsoniter.Unmarshal(invokeOutput.Payload, &engineOutput); err != nil {
+		return engineOutput, errors.Wrap(err, "failed to unmarshal engine response into PolicyEngineOutput")
+	}
```

Comment:
Our `genericapi` package provides an `Invoke` function which handles these 15 lines: marshaling the input, sending the request, checking for errors, and unmarshaling the response:

```suggestion
	err := genericapi.Invoke(e.lambdaClient, e.lambdaName, &input, &output)
```

Reply:
I know you were just copying the code, but might as well clean it up while we're here :)

Reply:
Of course, fixed